### PR TITLE
Add label parser support

### PR DESCRIPTION
**Pull request by AlanFoster**
closes https://github.com/tenderlove/aarch64/issues/400

This approach runs the parser in two phases:
- First pass - Collecting the required labels and offsets
- Second pass - Generating the instructions using the above information

### CODE REVIEW

File: lib/aarch64/parser.rb [hunk 1]
```diff
@@ -218,10 +611,20 @@ def _next_token
         end
       elsif str = @scan.scan(SYS_REG_SCAN)
         [:SYSTEMREG, SYS_REG_MAP[str.downcase]]
-      elsif str = @scan.scan(/\w+/)
+      elsif str = @scan.scan(KEYWORDS_SCAN)
```
Comment (AlanFoster) [inline lib/aarch64/parser.rb:614]: Falling through to parsing `/\w+` as keywords wasn't working for the added requirement of parsing labels; I swapped to following most of the patterns in the 'very basic lexer' example [here](https://tenderlovemaking.com/2023/09/02/fast-tokenizers-with-stringscanner.html) as the first step to getting the tests passing :+1:

File: lib/aarch64/parser.rb [hunk 2]
```diff
@@ -218,10 +611,20 @@ def _next_token
...
         [str.upcase.to_sym, str]
+      elsif str = @scan.scan(LABEL_SCAN)
+        [:LABEL, str]
       else
+        [:UNKNOWN_CHAR, @scan.getch]
       end
     end
+
+    def on_error(token_id, val, vstack)
```
Comment (AlanFoster) [inline lib/aarch64/parser.rb:623]: I found it hard to track down the parsing errors in the original input string when I was first playing with the gem; so I've tried to enhance this while I'm here. I can remove this change entirely or put it into a separate PR

File: lib/aarch64.rb [hunk 3]
```diff
@@ -288,6 +288,7 @@ def adr xd, label
     end
 
     def adrp xd, label
+      label = Immediate.new(label) if label.integer?
```
Comment (AlanFoster) [inline lib/aarch64.rb:291]: Aligning with `adr` above to ensure that the implementation can handle immediate values and labels

File: lib/aarch64/parser.y [hunk 4]
```diff
@@ -1057,6 +1068,21 @@ rule
       }
     ;
 
+  reg_imm_imm_or_label
```
Comment (AlanFoster) [inline lib/aarch64/parser.y:1071]: I'm intending that the user reads the last two values as being grouped, i.e.`reg_imm_imm_or_label ` should really parsed as `reg_imm_(imm_or_label)`. Happy to follow an alternative naming convention

Comment (tenderlove) [reply lib/aarch64/parser.y:1052]: I think this is fine for now.

**Review by tenderlove — CHANGES_REQUESTED** [review]
Do you mind removing the `.gitignore` file? Other than that, I think this patch is fine.

**Comment by AlanFoster** [comment]
I've removed that now :+1:

**Comment by AlanFoster** [comment]
Is there anything else I can do to help with this PR or https://github.com/tenderlove/aarch64/pull/403
I'm more than happy to make any other changes that are needed 💯

**Comment by tenderlove** [comment]
@AlanFoster sorry, I just hadn't been paying attention. Everything looks good here, I'll merge (and check #403)

**Comment by tenderlove** [comment]
@AlanFoster I'll ship a new version soon. I think it should be a 2.1.0 since we're mucking with dependencies.

**Pull request merged**